### PR TITLE
[RTC-405] Resolve FQDNs in DNS cluster strategy

### PR DESCRIPTION
**Pull request by mickel8**
After this PR:
* a default node name is `jellyfish@(hostname)` - thanks to this, when using DNS strategy in docker we don't have to set JF_DIST_NODE_NAME explicitly 
* if hostname in node name in DNS strategy is not an ip address, we will try to resolve it under the hood
* JF_DIST_MODE is snam by default - that's because of default node name
* when JF_DIST_ENABLED is set, we always start the node in Elixir code. This means that if we make a release and we want to connect to already running node, we have to use RELEASE env vars i.e.

RELEASE_NODE=<JF_DIST_NODE_NAME(resolved in case of DNS)> RELEASE_COOKIE=<JF_DIST_COOKIE> RELEASE_DISTRIBUTION=<JF_DIST_MODE> _build/../jellyfish remote 

To be precise, not always all of the above ones but this is quite tricky to describe 

Documentation update jellyfish-dev/jellyfish-docs#72

## Acknowledging the stipulations set forth:
 - [x] I hereby confirm that a Pull Request involving updates to the Software Development Kit (SDK) has been smoothly merged, currently awaits processing, or is otherwise deemed unnecessary in this context.
 - [x] I also affirm that another Pull Request, specifically addressing updates to the documentation body (commonly referred to as 'docs'), has either been successfully incorporated, is in the process of review, or is considered superfluous under the prevailing circumstances.

### CODE REVIEW

File: test/jellyfish/config_reader_test.exs [hunk 1]
```diff
@@ -129,37 +129,52 @@ defmodule Jellyfish.ConfigReaderTest do
...
+      System.put_env("JF_DIST_NODE_NAME", "testnodename")
+      assert_raise RuntimeError, fn -> ConfigReader.read_dist_config() end
+      System.delete_env("JF_DIST_NODE_NAME")
+      assert ConfigReader.read_dist_config()
+      System.put_env("JF_DIST_MODE", "invalid")
+      assert_raise RuntimeError, fn -> ConfigReader.read_dist_config() end
+
+      System.put_env("JF_DIST_MODE", "name")
+      System.put_env("JF_DIST_COOKIE", "testcookie")
+      System.put_env("JF_DIST_NODE_NAME", "testnodename@127.0.0.1")
```
Comment (Rados13) [inline test/jellyfish/config_reader_test.exs:172]: We should also check the situation in which we applied some of the required ENVs. Currently, this test checks only when we don't have set any ENVs and then that we set all of them.

File: lib/jellyfish/config_reader.ex [hunk 2]
```diff
@@ -230,4 +254,30 @@ defmodule Jellyfish.ConfigReader do
...
+  defp parse_mode("name"), do: :longnames
+  defp parse_mode("sname"), do: :shortnames
+  defp parse_mode(other), do: raise("Invalid JF_DIST_MODE. Expected sname or name, got: #{other}")
+
+  defp is_ip_address(hostname) do
+    case :inet.parse_address(String.to_charlist(hostname)) do
+      {:ok, _} -> true
+      {:error, _} -> false
+    end
+  end
```
[the remark this answers is not in the document]
Comment (mickel8) [reply lib/jellyfish/config_reader.ex:267]: It doesn't take the String as input :c

File: lib/jellyfish/config_reader.ex [hunk 3]
```diff
@@ -163,33 +171,49 @@ defmodule Jellyfish.ConfigReader do
...
+
+    node_name =
+      if is_ip_address(hostname) do
+        node_name
+      else
+        Logger.info(
+          "Resolving hostname part of JF node name as DNS cluster strategy requires IP address."
+        )
+
+        r_hostname = resolve_hostname(hostname)
```
[the remark this answers is not in the document]
Comment (mickel8) [reply lib/jellyfish/config_reader.ex:205]: Yeap, with the exception that this is searchable via CTRL+F, will change to `resolved_hostname`

File: test/jellyfish/config_reader_test.exs (hunk 1 above)
Comment (mickel8) [reply test/jellyfish/config_reader_test.exs:172]: Yeap as right now there are no required env vars :stuck_out_tongue:

File: lib/jellyfish/application.ex [hunk 4]
```diff
@@ -66,11 +66,11 @@ defmodule Jellyfish.Application do
   defp config_distribution(dist_config) do
     ensure_epmd_started!()
 
-    # Release always starts in a distributed mode
-    # so we have to start a node only in development.
-    # See env.sh.eex for more information.
+    # When running JF not in a cluster and using
+    # mix release, it starts in the distributed mode
+    # automatically
```
[the remark this answers is not in the document]
Comment (mickel8) [reply lib/jellyfish/application.ex:71]: by automatically I meant that by calling `./jellyfish start`. The start command runs `erl` with `sname` or `name` attributes starting BEAM as distributed node i.e. you can connect to it with some "random" cookie. When we start jellyfish with `JF_DIST_ENABLED=true` `./jellyfish start` will start BEAM as non-distributed node. We will change it into the distributed one when loading applicating, in Elixir code, using Node.start

File: rel/env.sh.eex [hunk 5]
```diff
@@ -3,18 +3,16 @@
...
+    # If Jellyfish is meant to be run in a cluster,
+    # leave node setup to the Elixir code where
+    # we do extra steps for DNS strategy to determine actual node name.
+    #
+    # We also try to read RELEASE_DISTRIBUTION env var
+    # to allow for calling `remote` command with JF_DIST_ENABLED set.
+    # In other case, RELEASE_DISTRIBUTION will always be set to true.
+    # This is a little hack. 
+    # We can get rid of it once we move to the reverse DNS strategy.
+    export RELEASE_DISTRIBUTION=${RELEASE_DISTRIBUTION:-none}
```
[the remark this answers is not in the document]
Comment (mickel8) [reply rel/env.sh.eex:15]: Yes, we will run in distributed mode but only to be able to call the `remote` command. Also, the cookie when `JF_DIST_ENABLED=false` will be different (it's some random string generated when calling `mix release`). I am not sure if we want to change this behavior.  We won't be able to connect to already running node which sometimes is quite useful feature. Even if we decide to do that I will make this in another PR

**Review by Rados13 — APPROVED** [review]

**Pull request merged**